Template Kotlin Refactoring, Convenience Features and Extensions

floschu opened the pull request:
* adds Snacker
* adds Toaster
* adds different kotlin extension classes
* adds more delegated properties
* refactors base view classes

floschu committed:
adds multiple convenience features and extensions

Commit:
adds multiple convenience features and extensions

floschu committed:
Merge branch 'template-kotlin-refactoring' of github.com:tailoredmedia/AndroidTemplate into template-kotlin-refactoring

migos commented on app/src/main/java/com/tailoredapps/template/ui/base/feedback/ActivityToaster.kt:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    private fun showInternal(title: String) {
+        toast = Toast.makeText(activity, title, Toast.LENGTH_LONG)
```
What is the reason you specify the lenght here? I think the user should have the possibility to specify it.

migos commented on app/src/main/java/com/tailoredapps/template/ui/base/feedback/ActivityToaster.kt:
```diff
@@ -0,0 +1,42 @@
+        toast?.show()
+    }
+
+    override fun hide() {
```
I would handle the hide inside the showInternal method

migos commented on app/src/main/java/com/tailoredapps/template/ui/base/feedback/Snacker.kt:
```diff
@@ -0,0 +1,33 @@
+import android.support.annotation.StringRes
+
+
+interface Snacker {
```
I would create two methods show(title: CharSequence) and showWithAction(title: CharSequence, action: () -> Unit, actionText: CharSequence) and remove the defaultActionText value. This would force the user to specify an actionText whenever an action is needed and minimize the risk of wrong behaviour (e.g. another dev changes the action but doesn't check what the current defaultActionText is)

migos commented on app/src/main/java/com/tailoredapps/template/ui/base/feedback/Snacker.kt:
```diff
@@ -0,0 +1,33 @@
+    fun show(@StringRes titleRes: Int, action: (() -> Unit)? = null, actionText: CharSequence? = null)
+    fun show(@StringRes titleRes: Int, action: (() -> Unit)? = null, @StringRes actionText: Int? = null)
+
+    fun hideActionSnack()
```
What is the reason to distinguish between hiding a snackbar with and without action when only one snackbar is visible at the same time? I would merge both methods into hideCurrentlyVisibleSnack() or handle it again inside the show method as mentioned in the review comment for the Toaster

migos commented on app/src/main/java/com/tailoredapps/template/ui/base/BaseActivity.kt:
```diff
@@ -94,30 +95,11 @@ abstract class BaseActivity<B : ViewDataBinding, VM : MvvmViewModel<*>> : AppCom
-            }
-        }
-
+        viewModel.attachViewOrThrow(this, savedInstanceState)
```
I would rename this method to attachViewOrThrowRuntimeException(...) for better readability

migos commented on app/src/main/java/com/tailoredapps/template/ui/base/feedback/Snacker.kt:
```diff
@@ -0,0 +1,33 @@
+    val activity: Activity
+    val defaultActionText: CharSequence
+
+    fun show(title: CharSequence, action: (() -> Unit)? = null, actionText: CharSequence? = null)
```
Let the dev specify the length

floschu committed:
implemented pull request feedback from @migos

floschu committed:
Merge branch 'template-kotlin' into template-kotlin-refactoring

# Conflicts:
#	app/src/main/java/com/tailoredapps/template/injection/components/ActivityComponent.kt
#	app/src/main/java/com/tailoredapps/template/ui/base/BaseActivityViewHolder.kt
#	app/src/main/java/com/tailoredapps/template/ui/base/BaseFragment.kt

floschu committed:
added useful realm extensions

floschu committed:
* changed navigator showdialogfragment method to be able to add a fragment tag
* added finishAffinity to navigator

floschu committed:
added appbar srolling view behaviour to main activity layout

floschu committed:
refactored toaster and snacker

floschu committed:
recatored activitysnacker and activitytoaster

floschu committed:
toaster is now per application

floschu committed:
fixed
*ActivitySnacker
*Different Extension classes
*Different Base classes

added
* RxDialogExtensions

deleted
* DIfferent unnecessary classes

patloew committed:
Merge branch 'template-kotlin' into template-kotlin-refactoring
